django/thunderstore/account/tests/test_user_flags.py

```diff
@@ -0,0 +1,166 @@
+
+
+@pytest.mark.django_db
+def test_user_flags_empty_for_unauthenticated(
+    user_flag: UserFlag,
+):
+    start_time = timezone.now()
+    late_time = timezone.now() + timedelta(days=10)
+
+    UserFlagMembershipFactory(
+        flag=user_flag,
+        datetime_valid_from=start_time,
+        datetime_valid_until=None,
+    )
+
+    assert UserFlag.get_active_flags_on_user(None, late_time) == []
+    assert UserFlag.get_active_flags_on_user(AnonymousUser(), late_time) == []
+
+
+@pytest.mark.django_db
+def test_user_flags_overlapping_memberships_return_distinct(
+    user: UserType,
+    user_flag: UserFlag,
+):
+    early_time = timezone.now()
+    start_time = early_time + timedelta(seconds=1)
+    late_time = start_time + timedelta(days=100000)
+
+    for _ in range(3):
+        UserFlagMembershipFactory(
+            user=user,
+            flag=user_flag,
+            datetime_valid_from=start_time,
+            datetime_valid_until=None,
+        )
+
+    assert UserFlag.get_active_flags_on_user(user, early_time) == []
+    assert UserFlag.get_active_flags_on_user(user, late_time) == [user_flag.identifier]
+    assert UserFlag.get_active_flags_on_user(None, late_time) == []
+
+
+@pytest.mark.django_db
+def test_user_flags_without_end_time_are_active(
+    user: UserType,
+    user_flag: UserFlag,
+):
+    early_time = timezone.now()
+    start_time = early_time + timedelta(seconds=1)
+    late_time = start_time + timedelta(days=100000)
+
+    UserFlagMembershipFactory(
+        user=user,
+        flag=user_flag,
+        datetime_valid_from=start_time,
+        datetime_valid_until=None,
+    )
+    assert UserFlag.get_active_flags_on_user(user, early_time) == []
+    assert UserFlag.get_active_flags_on_user(user, late_time) == [user_flag.identifier]
+    assert UserFlag.get_active_flags_on_user(None, late_time) == []
+
+
+@pytest.mark.django_db
+def test_user_flags_past_end_time_are_inactive(
+    user: UserType,
+    user_flag: UserFlag,
+):
+    early_time = timezone.now()
+    start_time = early_time + timedelta(days=1)
+    middle_time = start_time + timedelta(days=1)
+    almost_end_time = middle_time + timedelta(days=1)
+    end_time = almost_end_time + timedelta(microseconds=1)
+    late_time = end_time + timedelta(days=1)
+
+    UserFlagMembershipFactory(
+        user=user,
+        flag=user_flag,
+        datetime_valid_from=start_time,
+        datetime_valid_until=end_time,
+    )
+
+    assert UserFlag.get_active_flags_on_user(user, early_time) == []
+    assert UserFlag.get_active_flags_on_user(user, start_time) == [user_flag.identifier]
+    assert UserFlag.get_active_flags_on_user(user, middle_time) == [
+        user_flag.identifier
+    ]
+    assert UserFlag.get_active_flags_on_user(None, middle_time) == []
+    assert UserFlag.get_active_flags_on_user(user, almost_end_time) == [
+        user_flag.identifier
+    ]
+    assert UserFlag.get_active_flags_on_user(user, end_time) == []
```

Comment:
Not sure about this since I'm not native speaker, but should "valid until" be inclusive rather than exclusive limit?

Reply:
I could see this going either way, but via an example if you had a reservation until 12:00 for example, I'd read it as you having to have left the premises by 12:00 rather also including the first minute of 12:00 in it.

Anyway in this case it makes very little difference since we're talking about microseconds, I'm just used to having the start of time periods be inclusive and end not as that avoids creating unintended overlap with multiple periods.